### PR TITLE
pub package type local repository

### LINKED ISSUE #408: Missing Package Type Pub of Local Repository
**Describe the feature**
Can't create local repository of Pub package type

**Solution I'd like**
Please provide a template like other local package types which will enable user to create Pub type local repo

### PR DESCRIPTION
**Pull request by dasmanas**
closes #408 

- Package Type Pub added under repository types
  - Local
  - Remote
  - Virtual
- Remote repository test case has been enhanced with client SSL / TLS Certificate validation

### CODE REVIEW

**Review by alexhung — CHANGES_REQUESTED** [review]
`pub` repo type is also missing for remote and virtual repo categories. Let's add that as well.

**Review by maheshjfrog — COMMENTED** [review]
Looks ok. Like Alex suggested ... i think, we can add remote and virtual repo for this package type too.

File: docs/resources/artifactory_local_pub_repository.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+## Argument Reference
+
+Arguments have a one to one mapping with the [JFrog API](https://www.jfrog.com/confluence/display/RTF/Repository+Configuration+JSON). The following arguments are supported:
+
+* `key` - (Required) - the identity key of the repo
+* `description` - (Optional)
+* `notes` - (Optional)
+
+Arguments for Pub repository type closely match with arguments for Generic repository type.
```
Comment (danielmkn) [inline docs/resources/artifactory_local_pub_repository.md:21]: @alexhung should we start adding import instructions to repo docs?

Comment (alexhung) [reply docs/resources/artifactory_local_pub_repository.md:21]: @danielmkn good catch. Yes!

**Comment by dasmanas** [comment]
recheck

File: pkg/artifactory/resource_artifactory_remote_repository_test.go [hunk 2]
```diff
@@ -547,6 +547,8 @@ func TestAccRemoteRepository_nugetNew(t *testing.T) {
 func mkNewRemoteTestCase(repoType string, t *testing.T, extraFields map[string]interface{}) (*testing.T, resource.TestCase) {
 	_, fqrn, name := mkNames("terraform-remote-test-repo-full", fmt.Sprintf("artifactory_remote_%s_repository", repoType))
 
+	certificateAlias := "certificate" + randSeq(10)
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_remote_repository_test.go:550]: Is `randSeq()` necessary? Would `randomInt()` suffice?

File: pkg/artifactory/util_test.go [hunk 3]
```diff
@@ -2,13 +2,15 @@ package artifactory
...
 	"fmt"
-	"gopkg.in/yaml.v2"
+	"io/ioutil"
 	"math"
 	"net/http"
 	"reflect"
 	"strings"
 	"testing"
 
+	"gopkg.in/yaml.v2"
```
Comment (alexhung) [inline pkg/artifactory/util_test.go:12]: Please sort the import statements

File: sample.tf [hunk 4]
```diff
@@ -129,6 +129,11 @@ resource "artifactory_local_opkg_repository" "opkg-local" {
   description = "Repo created by Terraform Provider Artifactory"
 }
 
+resource "artifactory_local_pub_repository" "pub-local" {
+  key         = "pub-local"
+  description = "Repo created by Terraform Provider Artifactory"
+}
+
```
Comment (alexhung) [inline sample.tf:136]: Please add remote and virtual examples.

File: pkg/artifactory/util_test.go [hunk 5]
```diff
@@ -245,3 +247,30 @@ var deleteProxy = func(t *testing.T, proxyKey string) {
...
+
+	certFileBytes, err := ioutil.ReadFile("../../samples/cert.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = restyClient.R().
+		SetBody(string(certFileBytes)).
+		SetContentLength(true).
+		Post(fmt.Sprintf("%s%s", endpoint, certificateAlias))
```
Comment (alexhung) [inline pkg/artifactory/util_test.go:262]: Let's rename `endpoint` to a descriptive name.

**Review by alexhung — CHANGES_REQUESTED** [review]
A few changes required, and add import documentation to repos.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,12 @@
+## 6.5.0 (Apr 15, 2022). Tested on Artifactory 7.37.13
+
+IMPROVEMENTS:
+
+* Resources added for Pub package type of Local Repository [GH-409]
```
Comment (chb0github) [inline CHANGELOG.md:5]: Make sure this ticket number is linkable in the readme

File: docs/resources/artifactory_remote_pub_repository.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+```
+
+## Argument Reference
+
+Arguments have a one to one mapping with the [JFrog API](https://www.jfrog.com/confluence/display/RTF/Repository+Configuration+JSON). The following arguments are supported:
+
+* `key` - (Required) The repository identifier. Must be unique system-wide
+* `description` - (Optional)
+* `notes` - (Optional)
+* `url` - (Required) - the remote repo URL. You kinda don't have a remote repo without it
```
Comment (chb0github) [inline docs/resources/artifactory_remote_pub_repository.md:24]: The cheeky commentary is unnecessary - and yes, I originally put it in the others

**Review by maheshjfrog — COMMENTED** [review]
Same points as Alex's

File: pkg/artifactory/util.go [hunk 8]
```diff
@@ -116,6 +116,17 @@ func getMD5Hash(o interface{}) string {
 	return hex.EncodeToString(hasher.Sum(nil))
 }
 
+var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
```
Comment (chb0github) [inline pkg/artifactory/util.go:119]: So, this function should really be called `randomString` and please move the selectable characters into function itself. Also note: you can actually do this numerically the letter. Capital `A` is hex `0x40` - to get `B` you simply add 1

These sorts of library functions have to be out there in the wild

**Review by chb0github — CHANGES_REQUESTED** [review]
see comments

File: pkg/artifactory/resource_artifactory_remote_repository_test.go (hunk 2 above)
Comment (dasmanas) [reply pkg/artifactory/resource_artifactory_remote_repository_test.go:550]: changed to randomInt

File: pkg/artifactory/util_test.go (hunk 3 above)
Comment (dasmanas) [reply pkg/artifactory/util_test.go:12]:
```
import (
	"fmt"
	"io/ioutil"
	"math"
	"net/http"
	"reflect"
	"strings"
	"testing"

	"gopkg.in/yaml.v2"

	"github.com/go-resty/resty/v2"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
)
```
This is how it looks like now. As per go fmt it is soreted.

File: sample.tf (hunk 4 above)
Comment (dasmanas) [reply sample.tf:136]: done

File: pkg/artifactory/util_test.go (hunk 5 above)
Comment (dasmanas) [reply pkg/artifactory/util_test.go:262]: Was reusing the same variable already defined in `resource_artifactory_certificate.go` file. Changed the variable name to `certificateEndpoint`

File: pkg/artifactory/util.go (hunk 8 above)
Comment (dasmanas) [reply pkg/artifactory/util.go:119]: removed the method `randomSeq`. Instead use `randomInt` as per @alexhung suggestion.

File: CHANGELOG.md (hunk 6 above)
Comment (dasmanas) [reply CHANGELOG.md]: are you proposing like following.

[GH-409](https://github.com/jfrog/terraform-provider-artifactory/pull/409)

or #408  https://github.com/jfrog/terraform-provider-artifactory/issues/408

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,16 @@
+## 6.5.0 (Apr 15, 2022). Tested on Artifactory 7.37.14
+
+IMPROVEMENTS:
+
+* Resources added for Pub package type of Local Repository
+* Resources added for Pub package type of Remote Repository
+* Resources added for Pub package type of Virtual Repository
+* Acceptance test case enhanced with Client TLS Certificate
+
+[GH-409](https://github.com/jfrog/terraform-provider-artifactory/pull/409)
```
Comment (alexhung) [inline CHANGELOG.md:10]: Only link to PR is needed.

File: docs/resources/artifactory_local_pub_repository.md (hunk 1 above)
Comment (alexhung) [reply docs/resources/artifactory_local_pub_repository.md:21]: Check out documentation in other repositories. Most of them have import command example at the end of the file.

File: pkg/artifactory/util_test.go (hunk 3 above)
Comment (alexhung) [reply pkg/artifactory/util_test.go:12]: Run `make fmt` please. That uses `goimports` which has a different set of sorting rules.

File: sample.tf [hunk 10]
```diff
@@ -371,6 +376,11 @@ resource "artifactory_remote_p2_repository" "my-remote-p2" {
   url = "http://testartifactory.io/artifactory/example-p2/"
 }
 
+resource "artifactory_remote_pub_repository" "my-remote-pub" {
+  key         = "my-remote-pub"
```
Comment (alexhung) [inline sample.tf:380]: Please align the HCL

**Review by alexhung — CHANGES_REQUESTED** [review]

File: docs/resources/artifactory_local_pub_repository.md (hunk 1 above)
Comment (dasmanas) [reply docs/resources/artifactory_local_pub_repository.md:21]: done

File: CHANGELOG.md (hunk 9 above)
Comment (dasmanas) [reply CHANGELOG.md:10]: done

Comment (dasmanas) [reply CHANGELOG.md:10]: done

File: CHANGELOG.md (hunk 6 above)
Comment (dasmanas) [reply CHANGELOG.md]: done

File: sample.tf (hunk 10 above)
Comment (dasmanas) [reply sample.tf]: done

File: pkg/artifactory/util_test.go (hunk 3 above)
Comment (dasmanas) [reply pkg/artifactory/util_test.go:12]: after running `make fmt` on my workspace, following files are getting modified. `util_test.go` is not in the list

```
modified:   pkg/artifactory/datasource_artifactory_file.go
        modified:   pkg/artifactory/resource_artifactory_anonymous_user.go
        modified:   pkg/artifactory/resource_artifactory_anonymous_user_test.go
        modified:   pkg/artifactory/resource_artifactory_backup.go
        modified:   pkg/artifactory/resource_artifactory_managed_user_test.go
        modified:   pkg/artifactory/resource_artifactory_pull_replication.go
        modified:   pkg/artifactory/resource_artifactory_replication_config.go
        modified:   pkg/artifactory/resource_artifactory_single_replication_config.go
        modified:   pkg/artifactory/resource_artifactory_single_replication_config_test.go
        modified:   pkg/artifactory/resource_artifactory_user.go
        modified:   pkg/artifactory/resource_artifactory_virtual_debian_repository.go
        modified:   pkg/artifactory/resource_artifactory_webhook.go
        modified:   pkg/artifactory/resource_artifactory_webhook_base.go
        modified:   pkg/artifactory/user.go
```

Comment (alexhung) [reply pkg/artifactory/util_test.go:12]: Ok, the other import statements in the code have `gopkg.in/yaml.v2` in after `github.com` for obvious reason. Please make this consistent.

**Review by alexhung — CHANGES_REQUESTED** [review]
Couple of outstanding changes

File: docs/resources/artifactory_remote_pub_repository.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
```
Comment (chb0github) [inline docs/resources/artifactory_remote_pub_repository.md:24]:
```suggestion
* `url` - (Required) - the remote repo URL. 
```

**Comment by dasmanas** [comment]
> Looks ok. Like Alex suggested ... i think, we can add remote and virtual repo for this package type too.

Pub added under remote and virtual both

**Comment by dasmanas** [comment]
Moved all changes under https://github.com/jfrog/terraform-provider-artifactory/pull/421.

**Pull request closed without merging**